### PR TITLE
[python][addons] Request setResolvedUrl callback even when previously resolved

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

Allow Kodi to request setResolvedUrl callback also if the addon listitem was previously resolved

## Motivation and Context

When we play addon videos from playlist, and we try to play a video previously played
Kodi avoid to call again the addon to resolve the ListItem,
this is correct, but at today the video addons use this callback not only to provide the ListItem to play with `xbmcplugin.setResolvedUrl` but also to execute code (needed before or after Kodi full execute the video) or/and to keep track of some things when a user play a video.

This is to solve issue #19529
if it's good i ask for backport permission

## How Has This Been Tested?

On the add-on play callback:

```python
list_item = xbmcgui.ListItem(offscreen=True)
...
list_item.setProperty('ForceResolvePlugin', 'true')
xbmcplugin.setResolvedUrl(PLUGIN_HANDLE, succeeded=True, listitem=list_item)
```

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by lrusak** [comment]
jenkins build this please

**Comment by enen92** [comment]
This will likely break nested plugin items, eg a plugin that resolves it's dynpath to plugin://video.youtube

**Comment by CastagnaIT** [comment]
you know an addon that make the use of nested plugin so i can try to test?

**Comment by MoojMidge** [comment]
@CastagnaIT example plugin that resolves to itself a number of times is attached for testing: [plugin.video.example.zip](https://github.com/xbmc/xbmc/files/6408905/plugin.video.example.zip). Can change the number of attempts on line 222 of main.py

As far as I can see the code changed in this PR, in PlayFile(), just makes subsequent attempts to resolve a listItem the same as the very first attempt, so shouldn't break anything.

However the same code is also in PlayMedia(): https://github.com/xbmc/xbmc/blob/ce02a174e1d77583f51a6113968b6336dae68817/xbmc/Application.cpp#L2578-L2585

Would need to change both locations to fix for playlists and plugin directory listing. Note that similar/same code is also used elsewhere in Kodi, but I'm not sure of the implication of changing this elsewhere.

**Comment by enen92** [comment]
@CastagnaIT maybe plugin.video.nasa from the repo might work for testing

**Comment by CastagnaIT** [comment]
I had to add also this
```c++
  if (!URIUtils::IsPlugin(itemPath) && URIUtils::IsPlugin(item.GetDynPath()))
    itemPath = item.GetDynPath();
```
because when you try to play eg. a STRM file from library, the Path will be the real filename and in the DynPath there is the plugin path

@MoojMidge the mentioned code in CApplication::PlayMedia is never used in our cases, seem used mainly by PVR live videos and games, for single run only, in fact if you run a game it will be executed,
i made the same changes anyway, otherwise it would have been inconsistent,
but i can not test PVR live, if someone can do a fast test to ensure there is no regression for completeness

@enen92 i can not use the plugin.video.nasa because use Tubed addon that is currently broken due to quotaExceeded unsolvable error, but from what i see the test plugin (thanks @MoojMidge) report in the log all recursive plugin call correctly

**Review by enen92 — COMMENTED** [review]
I still find this a bit hard to follow/read and it will be easy to break in the future :/ can you think of a better way of encapsulating the code?
I mean, the condition is basically "HasPluginPath", the itemPath is basically the lastResolvedPath. Maybe a new method will be harder to break/easy to follow instead of this

**Comment by MoojMidge** [comment]
If encapsulation is required, maybe the resolve loop should be moved into [GetPluginResult](https://github.com/xbmc/xbmc/blob/267090d82a2d04043eb7931495ad991f393accd9/xbmc/filesystem/PluginDirectory.cpp#L73-L99) instead of calling GetPluginResult from within a loop?

Something like:

```
bool CPluginDirectory::GetPluginResult(const std::string& strPath, CFileItem &resultItem, bool resume)
{
  // Override strPath with dynPath if strPath is not a plugin but dynPath is
  if (!URIUtils::IsPlugin(strPath) && URIUtils::IsPlugin(resultItem.GetDynPath()))
    strPath = resultItem.GetDynPath();

  for (int i = 0; URIUtils::IsPlugin(strPath) && i < 5; ++i)
  {
    CURL url(strPath);
    CPluginDirectory newDir;
  
    bool success = newDir.StartScript(strPath, resume);

    if (!success || resultItem.GetPath() == newDir.m_fileResult->GetPath())
    {
      success = false;
      break;
    }
  
    // update the play path and metadata, saving the old one as needed
    if (!resultItem.HasProperty("original_listitem_url"))
    resultItem.SetProperty("original_listitem_url", resultItem.GetPath());
    resultItem.SetDynPath(newDir.m_fileResult->GetPath());
    resultItem.SetMimeType(newDir.m_fileResult->GetMimeType());
    resultItem.SetContentLookup(newDir.m_fileResult->ContentLookup());
  
    if (resultItem.HasProperty("OverrideInfotag") && resultItem.GetProperty("OverrideInfotag").asBoolean())
      resultItem.UpdateInfo(*newDir.m_fileResult);
    else
      resultItem.MergeInfo(*newDir.m_fileResult);
  
    if (newDir.m_fileResult->HasVideoInfoTag() && newDir.m_fileResult->GetVideoInfoTag()->GetResumePoint().IsSet())
    {
      resultItem.m_lStartOffset = STARTOFFSET_RESUME; // resume point set in the resume item, so force resume
      resume = true;
    }
    else
      resume = false;

    strPath = resultItem.GetDynPath();
  }

  return success;
}
```


The just replace the whole resolve loop with:
```
if !XFILE::CPluginDirectory::GetPluginResult(item.GetPath(), item, item.m_lStartOffset == STARTOFFSET_RESUME)
  return false;
```

https://github.com/xbmc/xbmc/search?q=GetPluginResult

**Comment by CastagnaIT** [comment]
@enen92 i have tried do better, when you have time give a look

**Comment by enen92** [comment]
Looks better. I am wondering if it is possible to only follow this behaviour if a specific property is set on the item. Avoiding cache and forcing Kodi to resolve an already resolved path will incur in a performance penalty due to yet another cpp-python call. Ideally, we should be able to specify this behaviour on the plugin side.

**Comment by CastagnaIT** [comment]
I will give it a try

**Comment by CastagnaIT** [comment]
I have done i have called the property as "EnsurePathCallback"
do you think that there is possibility to have this fix/improvement also on Matrix?

File: xbmc/FileItem.cpp [hunk 1]
```diff
@@ -1933,6 +1933,15 @@ void CFileItem::SetDynPath(const std::string &path)
   m_strDynPath = path;
 }
 
+const std::string& CFileItem::GetPluginPath() const
```
Comment (enen92) [inline xbmc/FileItem.cpp:1936]: I don't like the name of this method. Dynpath always take precedence over path, the resolve instruction just takes whatever is in path and resolves it to dynpath. However, this method might return the original path. Maybe something like GetOriginalPluginPath. Also I don't think this belongs to CFileItem as it is specific to an item type (plugin). Can you move this to URIUtils or a similar class? Anything that avoids stuffing new methods in CFileItem is better.

File: xbmc/FileItem.h [hunk 2]
```diff
@@ -135,6 +135,7 @@ class CFileItem :
   void SetDynURL(const CURL& url);
   const std::string &GetDynPath() const;
   void SetDynPath(const std::string &path);
+  const std::string& GetPluginPath() const;
```
Comment (enen92) [inline xbmc/FileItem.h:138]: Missing docs

File: xbmc/filesystem/PluginDirectory.cpp [hunk 3]
```diff
@@ -70,6 +70,42 @@ bool CPluginDirectory::StartScript(const std::string& strPath, bool resume)
...
+    lastResolvedPath = resultItem.GetPluginPath();
+  }
+  else
+    lastResolvedPath = resultItem.GetDynPath();
+
+  if (!lastResolvedPath.empty())
+  {
+    // we try to resolve recursively up to 5 nested plugin paths to avoid deadlocks
+    //  (plugin:// paths can resolve to plugin:// paths)
+    for (int i = 0; URIUtils::IsPlugin(lastResolvedPath) && i < 5; ++i)
```
Comment (enen92) [inline xbmc/filesystem/PluginDirectory.cpp:90]: Since the 5 magic number is now restricted to this class, can you move it to a const value in this file? Anonymous namespace

File: xbmc/filesystem/PluginDirectory.h [hunk 4]
```diff
@@ -34,6 +34,13 @@ class CPluginDirectory : public IDirectory, public CRunningScriptsHandler<CPlugi
   float GetProgress() const override;
   void CancelDirectory() override;
   static bool RunScriptWithParams(const std::string& strPath, bool resume);
+
+  /*! \brief Get a reproducible CFileItem by trying to resolve the plugin paths up to 5 times.
```
Comment (enen92) [inline xbmc/filesystem/PluginDirectory.h:38]: Please remove the reference to 5 and use max allowed nested resolutions or similar

File: xbmc/interfaces/legacy/ListItem.h [hunk 5]
```diff
@@ -996,9 +996,11 @@ namespace XBMCAddon
       /// | StationName   | string ("My Station Name") - Used to enforce/override MusicPlayer.StationName infolabel from addons (e.g. in radio addons)
       /// | TotalTime     | float (7848.0) - Set the total time of the item in seconds
       /// | OverrideInfotag | string - "true", "false" - When true will override all info from previous listitem
+      /// | EnsurePathCallback | string - "true", "false" - When true ensures that a plugin will always receive callbacks to resolve paths (useful for playlist cases)
```
Comment (enen92) [inline xbmc/interfaces/legacy/ListItem.h:999]: Maybe `ForceResolvePlugin`?

**Review by enen92 — CHANGES_REQUESTED** [review]
Looks far better and I am finally okay with it. Can you take my comments into consideration?

**Comment by enen92** [comment]
Regarding a possible backport I am fine with it as the new behaviour will only apply if that property is set.

File: xbmc/FileItem.cpp (hunk 1 above)
Comment (CastagnaIT) [reply xbmc/FileItem.cpp]: done

File: xbmc/FileItem.h (hunk 2 above)
Comment (CastagnaIT) [reply xbmc/FileItem.h]: done

File: xbmc/filesystem/PluginDirectory.cpp (hunk 3 above)
Comment (CastagnaIT) [reply xbmc/filesystem/PluginDirectory.cpp]: done, take a look on this if is correct

File: xbmc/filesystem/PluginDirectory.h (hunk 4 above)
Comment (CastagnaIT) [reply xbmc/filesystem/PluginDirectory.h]: done

File: xbmc/interfaces/legacy/ListItem.h (hunk 5 above)
Comment (CastagnaIT) [reply xbmc/interfaces/legacy/ListItem.h]: done

**Review by enen92 — APPROVED** [review]

**Comment by enen92** [comment]
@CastagnaIT check build failures. Maybe the best location for the new method is a public one in CPluginDirectory to avoid adding the CFileItem dep/reference to URIUtils.

File: xbmc/interfaces/legacy/ListItem.h [hunk 6]
```diff
@@ -996,9 +996,11 @@ namespace XBMCAddon
       /// | StationName   | string ("My Station Name") - Used to enforce/override MusicPlayer.StationName infolabel from addons (e.g. in radio addons)
       /// | TotalTime     | float (7848.0) - Set the total time of the item in seconds
       /// | OverrideInfotag | string - "true", "false" - When true will override all info from previous listitem
+      /// | ForceResolvePlugin | string - "true", "false" - When true ensures that a plugin will always receive callbacks to resolve paths (useful for playlist cases)
       ///
       ///-----------------------------------------------------------------------
       /// @python_v20 OverrideInfotag property added
+      /// @python_v20 ForceResolvePlugin property added
```
Comment (CastagnaIT) [inline xbmc/interfaces/legacy/ListItem.h:1003]: but if i backport the pr the "v20" note would be inconsistent...i do not backport

File: xbmc/Application.cpp [hunk 7]
```diff
@@ -2582,17 +2582,8 @@ void CApplication::Stop(int exitCode)
...
-    bool resume = item.m_lStartOffset == STARTOFFSET_RESUME;
-
-    if (!XFILE::CPluginDirectory::GetPluginResult(item.GetDynPath(), item, resume) ||
-        item.GetDynPath() == item.GetPath()) // GetPluginResult resolved to an empty path
-      return false;
-  }
-  // if after the 5 resolution attempts the item is still a plugin just return, it isn't playable
-  if (URIUtils::IsPlugin(item.GetDynPath()))
+  // if the item is a plugin we need to resolve the plugin paths
+  if (!XFILE::CPluginDirectory::GetResolvedPluginResult(item))
```
Comment (enen92) [inline xbmc/Application.cpp:2586]: @CastagnaIT sorry I missed this earlier and also for so many requests. I don't like the fact we are querying the plugin directory by default for each item that is requested for playback at a time we don't yet know the type of item. We should only enter this method if we know for sure at least one of the paths in the item is a `plugin://`.

Can you please add a new method to `URIUtils` called `HasPlugin` or `HasPluginPath`? This would return true if either `path` or `dynPath` is `plugin://`. Please then branch the `GetResolvedPluginResult` call only if `HasPluginPath` is true.
Thanks

Comment (CastagnaIT) [reply xbmc/Application.cpp:2586]: if i add `HasPluginPath` inside of `URIUtils ` will cause again jenkins error for the CFileItem reference
can i put the method in FileItem file?

Comment (enen92) [reply xbmc/Application.cpp:2586]: You could use a forward declarion for CFileItem on URIUtils.h, or just define the method in a way that it receives both paths directly

File: xbmc/filesystem/PluginDirectory.cpp [hunk 8]
```diff
@@ -28,6 +28,28 @@ using namespace XFILE;
...
+{
+const unsigned int maxPluginResolutions = 5;
+
+/*!
+  \brief Get the plugin path from a CFileItem.
+
+  \param item CFileItem where to get the path.
+  \return The plugin path if found otherwise an empty string.
+*/
+std::string GetOriginalPluginPath(CFileItem& item)
```
Comment (enen92) [inline xbmc/filesystem/PluginDirectory.cpp:41]:
```suggestion
const std::string GetOriginalPluginPath(CFileItem& item)
```

File: xbmc/filesystem/PluginDirectory.cpp [hunk 9]
```diff
@@ -70,6 +92,42 @@ bool CPluginDirectory::StartScript(const std::string& strPath, bool resume)
   return RunScript(this, addon, strPath, resume);
 }
 
+bool CPluginDirectory::GetResolvedPluginResult(CFileItem& resultItem)
```
Comment (enen92) [inline xbmc/filesystem/PluginDirectory.cpp:95]:
```suggestion
const bool CPluginDirectory::GetResolvedPluginResult(CFileItem& resultItem)
```

File: xbmc/filesystem/PluginDirectory.cpp [hunk 10]
```diff
@@ -70,6 +92,42 @@ bool CPluginDirectory::StartScript(const std::string& strPath, bool resume)
...
+    lastResolvedPath = GetOriginalPluginPath(resultItem);
+  }
+  else
+    lastResolvedPath = resultItem.GetDynPath();
+
+  if (!lastResolvedPath.empty())
+  {
+    // we try to resolve recursively up to n. (maxPluginResolutions) nested plugin paths
+    // to avoid deadlocks (plugin:// paths can resolve to plugin:// paths)
+    for (int i = 0; URIUtils::IsPlugin(lastResolvedPath) && i < maxPluginResolutions; ++i)
```
Comment (enen92) [inline xbmc/filesystem/PluginDirectory.cpp:112]:
```suggestion
    for (unsigned int i = 0; URIUtils::IsPlugin(lastResolvedPath) && i < maxPluginResolutions; ++i)
```

File: xbmc/filesystem/PluginDirectory.h [hunk 11]
```diff
@@ -34,6 +34,13 @@ class CPluginDirectory : public IDirectory, public CRunningScriptsHandler<CPlugi
   float GetProgress() const override;
   void CancelDirectory() override;
   static bool RunScriptWithParams(const std::string& strPath, bool resume);
+
+  /*! \brief Get a reproducible CFileItem by trying to recursively resolve the plugin paths
+  up to a maximum allowed limit. If no plugin paths exist it will be ignored.
+  \param resultItem the CFileItem with plugin paths to be resolved.
+  \return false if the plugin path cannot be resolved, true otherwise.
+  */
+  static bool GetResolvedPluginResult(CFileItem& resultItem);
```
Comment (enen92) [inline xbmc/filesystem/PluginDirectory.h:43]:
```suggestion
  const static bool GetResolvedPluginResult(CFileItem& resultItem);
```

File: xbmc/Application.cpp (hunk 7 above)
Comment (enen92) [reply xbmc/Application.cpp:2586]: Gave it a go, just forward declare CFIleItem on URIUtils.h, in URIUtils.cpp CFileItem is already included:

URIUtils.h:
```cpp
diff --git a/xbmc/utils/URIUtils.h b/xbmc/utils/URIUtils.h
index 8432c949e6..7e774143f3 100644
--- a/xbmc/utils/URIUtils.h
+++ b/xbmc/utils/URIUtils.h
@@ -13,6 +13,7 @@
 
 class CURL;
 class CAdvancedSettings;
+class CFileItem;
 
 class URIUtils
 {
@@ -28,6 +29,9 @@ public:
   static std::string GetExtension(const CURL& url);
   static std::string GetExtension(const std::string& strFileName);
 
+  // please add docs
+  const static bool HasPluginPath(const CFileItem& item);
```

```cpp
--- a/xbmc/utils/URIUtils.cpp
+++ b/xbmc/utils/URIUtils.cpp
@@ -66,6 +66,11 @@ std::string URIUtils::GetExtension(const std::string& strFileName)
   return strFileName.substr(period);
 }
 
+const bool URIUtils::HasPluginPath(const CFileItem& item)
+{
+  return IsPlugin(item.GetPath()) || IsPlugin(item.GetDynPath());
+}
+
```

Comment (enen92) [inline xbmc/Application.cpp:2586]: wrap around `HasPluginPath`

File: xbmc/Application.cpp [hunk 12]
```diff
@@ -2719,16 +2710,8 @@ bool CApplication::PlayFile(CFileItem item, const std::string& player, bool bRes
...
```
Comment (enen92) [inline xbmc/Application.cpp:2714]: wrap around `HasPluginPath`

File: xbmc/PlayListPlayer.cpp [hunk 13]
```diff
@@ -911,12 +911,8 @@ void PLAYLIST::CPlayListPlayer::OnApplicationMessage(KODI::MESSAGING::ThreadMess
         {
           CFileItemPtr item = (*list)[0];
           // if the item is a plugin we need to resolve the URL to ensure the infotags are filled.
-          // resolve only for a maximum of 5 times to avoid deadlocks (plugin:// paths can resolve to plugin:// paths)
-          for (int i = 0; URIUtils::IsPlugin(item->GetDynPath()) && i < 5; ++i)
-          {
-            if (!XFILE::CPluginDirectory::GetPluginResult(item->GetDynPath(), *item, true))
-              return;
-          }
+          if (!XFILE::CPluginDirectory::GetResolvedPluginResult(*item))
```
Comment (enen92) [inline xbmc/PlayListPlayer.cpp:914]: wrap around `HasPluginPath`

File: xbmc/filesystem/PluginDirectory.cpp [hunk 14]
```diff
@@ -70,6 +92,42 @@ bool CPluginDirectory::StartScript(const std::string& strPath, bool resume)
...
+      bool resume = resultItem.m_lStartOffset == STARTOFFSET_RESUME;
+
+      // we modify the item so that it becomes a real URL
+      if (!XFILE::CPluginDirectory::GetPluginResult(lastResolvedPath, resultItem, resume) ||
+          resultItem.GetDynPath() ==
+              resultItem.GetPath()) // GetPluginResult resolved to an empty path
+        return false;
+      lastResolvedPath = resultItem.GetDynPath();
+    }
+    // if after the 5 resolution attempts the item is still a plugin just return, it isn't playable
```
Comment (enen92) [inline xbmc/filesystem/PluginDirectory.cpp:123]:
```suggestion
    // if after the maximum allowed resolution attempts the item is still a plugin just return, it isn't playable
```

**Review by enen92 — CHANGES_REQUESTED** [review]
Last round I promise. Runtime tested, works well!

File: xbmc/PlayListPlayer.cpp [hunk 15]
```diff
@@ -911,12 +911,9 @@ void PLAYLIST::CPlayListPlayer::OnApplicationMessage(KODI::MESSAGING::ThreadMess
...
           // if the item is a plugin we need to resolve the URL to ensure the infotags are filled.
-          // resolve only for a maximum of 5 times to avoid deadlocks (plugin:// paths can resolve to plugin:// paths)
-          for (int i = 0; URIUtils::IsPlugin(item->GetDynPath()) && i < 5; ++i)
-          {
-            if (!XFILE::CPluginDirectory::GetPluginResult(item->GetDynPath(), *item, true))
-              return;
-          }
+          if (URIUtils::HasPluginPath(item) &&
+              !XFILE::CPluginDirectory::GetResolvedPluginResult(*item))
+            return;
```
Comment (howie-f) [inline xbmc/PlayListPlayer.cpp:916]:
```suggestion
          if (URIUtils::HasPluginPath(item) &&
              !XFILE::CPluginDirectory::GetResolvedPluginResult(*item))
          {
            return;
          }
```

File: xbmc/filesystem/PluginDirectory.cpp (hunk 8 above)
Comment (howie-f) [inline xbmc/filesystem/PluginDirectory.cpp:41]:
```suggestion
std::string GetOriginalPluginPath(const CFileItem& item)
```

File: xbmc/filesystem/PluginDirectory.cpp [hunk 16]
```diff
@@ -70,6 +92,42 @@ bool CPluginDirectory::StartScript(const std::string& strPath, bool resume)
...
+  std::string lastResolvedPath;
+  if (resultItem.HasProperty("ForceResolvePlugin") &&
+      resultItem.GetProperty("ForceResolvePlugin").asBoolean())
+  {
+    // ensures that a plugin have the callback to resolve the paths in any case
+    // also when the same items in the playlist are played more times
+    lastResolvedPath = GetOriginalPluginPath(resultItem);
+  }
+  else
+    lastResolvedPath = resultItem.GetDynPath();
```
Comment (howie-f) [inline xbmc/filesystem/PluginDirectory.cpp:106]:
```suggestion
  else
  {
    lastResolvedPath = resultItem.GetDynPath();
  }
```

File: xbmc/filesystem/PluginDirectory.cpp [hunk 17]
```diff
@@ -70,6 +92,42 @@ bool CPluginDirectory::StartScript(const std::string& strPath, bool resume)
...
+    // to avoid deadlocks (plugin:// paths can resolve to plugin:// paths)
+    for (unsigned int i = 0; URIUtils::IsPlugin(lastResolvedPath) && i < maxPluginResolutions; ++i)
+    {
+      bool resume = resultItem.m_lStartOffset == STARTOFFSET_RESUME;
+
+      // we modify the item so that it becomes a real URL
+      if (!XFILE::CPluginDirectory::GetPluginResult(lastResolvedPath, resultItem, resume) ||
+          resultItem.GetDynPath() ==
+              resultItem.GetPath()) // GetPluginResult resolved to an empty path
+        return false;
```
Comment (howie-f) [inline xbmc/filesystem/PluginDirectory.cpp:120]:
```suggestion
              resultItem.GetPath()) // GetPluginResult resolved to an empty path
      {
        return false;
      }
```

**Comment by enen92** [comment]
jenkins build this please

**Review by howie-f — COMMENTED** [review]
just some minor suggestions for readability.. would be nice if you could change it..
but it's not really mandatory. otherwise LGTM, thanks

File: xbmc/PlayListPlayer.cpp [hunk 18]
```diff
@@ -911,12 +911,9 @@ void PLAYLIST::CPlayListPlayer::OnApplicationMessage(KODI::MESSAGING::ThreadMess
...
           CFileItemPtr item = (*list)[0];
           // if the item is a plugin we need to resolve the URL to ensure the infotags are filled.
-          // resolve only for a maximum of 5 times to avoid deadlocks (plugin:// paths can resolve to plugin:// paths)
-          for (int i = 0; URIUtils::IsPlugin(item->GetDynPath()) && i < 5; ++i)
-          {
-            if (!XFILE::CPluginDirectory::GetPluginResult(item->GetDynPath(), *item, true))
-              return;
-          }
+          if (URIUtils::HasPluginPath(item) &&
+              !XFILE::CPluginDirectory::GetResolvedPluginResult(*item))
```
Comment (howie-f) [inline xbmc/PlayListPlayer.cpp:915]: you need to dereference both `item`s in order to get it build

```suggestion
          if (URIUtils::HasPluginPath(*item) &&
              !XFILE::CPluginDirectory::GetResolvedPluginResult(*item))
```

**Review by howie-f — CHANGES_REQUESTED** [review]

**Review by howie-f — APPROVED** [review]
if you add blank lines to `GetOriginalPluginPath` (right after the early returns) it‘s getting really pretty 🤩 last comment from my side

**Review by enen92 — APPROVED** [review]
Please rebase and it's good to go
